examples/seismic/tutorials/fwi_gradient_marmousi.py

```diff
@@ -0,0 +1,118 @@
+        diff_eq = Eq(
+            residual, dsyn.subs({dsyn.dimensions[-1]: residual.dimensions[-1]}) -
+            dobs.subs({dobs.dimensions[-1]: residual.dimensions[-1]})
+        )
+        Operator(diff_eq)()
+    else:
+        residual.data[:] = dsyn.data[:] - dobs.data[:]
+    return residual
+
+
+# FWI Gradient Kernel
+def fwi_gradient(mode, model, solver, geometry, source_locations, vp_in, factor=None):
+    grad = Function(name="grad", grid=model.grid)
+    residual = Receiver(name='residual', grid=model.grid,
+                        time_range=geometry.time_axis,
+                        coordinates=geometry.rec_positions)
+    d_obs = Receiver(name='d_obs', grid=model.grid,
+                     time_range=geometry.time_axis,
+                     coordinates=geometry.rec_positions)
+    d_syn = Receiver(name='d_syn', grid=model.grid,
+                     time_range=geometry.time_axis,
+                     coordinates=geometry.rec_positions)
+    objective = 0.0
+    for i in range(nshots):
+        geometry.src_positions[0, :] = source_locations[i, :]
+        solver.forward(vp=model.vp, rec=d_obs)
+        save_value = True if mode == "full" else False
+        _, u0, _ = solver.forward(vp=vp_in, save=save_value, rec=d_syn, factor=factor)
+
+        compute_residual(residual, d_obs, d_syn)
+        objective += 0.5 * norm(residual)**2
+
+        solver.gradient(rec=residual, u=u0, vp=vp_in, grad=grad, factor=factor)
+
+    return objective, grad
+
+
+def main(mode, factor=None):
+    shape = (601, 221)
+    spacing = (15.0, 15.0)
+    origin = (0.0, 0.0)
+    vel_path = '../../../devito/data/Marm.bin'
+    model_true = np.fromfile(vel_path, np.float32).reshape(221, 601)
+    msk = mask(model_true, 1.5)
+    model_init = gaussian_filter(model_true, sigma=[10, 15]) * msk
+    model_init[model_init == 0] = 1.5
+
+    model = Model(vp=model_true.T, origin=origin, shape=shape, spacing=spacing,
+                  space_order=2, nbl=20, bcs="damp")
+    model0 = Model(vp=model_init.T, origin=origin, shape=shape, spacing=spacing,
+                   space_order=2, nbl=20, bcs="damp", grid=model.grid)
+
+    t0, tn, f0 = 0.0, 4000.0, 0.005
+    src_coordinates = np.array([[model.domain_size[0] * 0.5, 20.0]])
+    rec_coordinates = np.column_stack((
+        np.linspace(0, model.domain_size[0], nreceivers),
+        np.full(nreceivers, 20.0)
+    ))
+    geometry = AcquisitionGeometry(model, rec_coordinates, src_coordinates,
+                                   t0, tn, f0=f0, src_type='Ricker')
+
+    solver = AcousticWaveSolver(model, geometry, space_order=4)
+
+    source_locations = np.column_stack((
+        np.linspace(0.0, model.domain_size[0], nshots),
+        np.zeros(nshots)
+    ))
+
+    ff, grad = fwi_gradient(mode, model, solver, geometry,
+                            source_locations, model0.vp, factor)
+    mem_usage = memory_usage()[0]
+    print(f"Memory usage at the end of gradient ({mode} mode): {mem_usage:.2f} MiB")
```

Comment:
Not sure what the goal of this check is here. the gradient is computed in as separate function so all the memory is freed. Showing memory usage difference would need to do this between forward and gradient for a single shot